src/OpenCensus.Abstractions/Trace/TraceState.cs

```diff
@@ -0,0 +1,301 @@
+                    this.entries = new List<Entry>(this.parent.Entries);
+                }
+
+                for (int i = 0; i < this.entries.Count; i++)
+                {
+                    if (this.entries[i].Key.Equals(key))
+                    {
+                        this.entries.RemoveAt(i);
+                        // Exit now because the entries list cannot contain duplicates.
+                        break;
+                    }
+                }
+
+                return this;
+            }
+
+            /// <summary>
+            /// Builds the tracestate.
+            /// </summary>
+            /// <returns>Resulting tracestate.</returns>
+            public Tracestate Build()
+            {
+                if (this.entries == null)
+                {
+                    return this.parent;
+
+                }
+
+                return Tracestate.Create(this.entries);
+            }
+        }
+
+        /// <summary>
+        /// Immutable tracestate entry.
+        /// </summary>
+        public sealed class Entry
+        {
+            private readonly string key;
+            private readonly string value;
+
+            private Entry(string key, string value)
+            {
+                this.key = key;
+                this.value = value;
+            }
+
+            /// <summary>
+            /// Gets the key of tracestate entry.
+            /// </summary>
+            public string Key { get => key; }
+
+            /// <summary>
+            /// Gets the value of tracestate entry.
+            /// </summary>
+            public string Value { get => value; }
+
+            /// <summary>
+            /// Creates a new Entry with the given name and value.
+            /// </summary>
+            /// <param name="key">Key of tracestate entry.</param>
+            /// <param name="value">Value of thacestate entry.</param>
+            /// <returns>The new tracestate entry.</returns>
+            public static Entry Create(string key, string value)
+            {
+                key = key ?? throw new ArgumentNullException(nameof(key));
+                value = value ?? throw new ArgumentNullException(nameof(value));
+
+                if (!ValidateKey(key))
+                {
+                    throw new ArgumentException("Doesn't comply to spec", nameof(key));
+                }
+
+                if (!ValidateValue(value))
+                {
+                    throw new ArgumentException("Doesn't comply to spec", nameof(value));
+                }
+
+                return new Entry(key, value);
+            }
+        }
+
+        private static bool ValidateKey(string key)
+        {
+            // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
+            // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+            // forward slashes /.
+
+            if (string.IsNullOrEmpty(key)
```

Comment:
what about '@'

Reply:
I hope that one of not passed tests in W3C check for this. Applied trivial fix for now